src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py

```diff
@@ -0,0 +1,248 @@
+            component_id: component id
+            metrics: metrics that should be fetched from this component.
+
+        Returns:
+            This class instance.
+        """
+        self: ComponentMetricFetcher = ComponentMetricFetcher.__new__(cls)
+        self._component_id = component_id
+        self._metrics = metrics
+        return self
+
+    @abstractmethod
+    async def fetch_next(self) -> Optional[ComponentMetricsData]:
+        """Fetch metrics for this component."""
+
+
+class LatestMetricsFetcher(ComponentMetricFetcher, Generic[T], ABC):
+    """Subscribe for the latest component data and extract the needed metrics."""
+
+    _receiver: Receiver[T]
+
+    @classmethod
+    async def async_new(
+        cls,
+        component_id: int,
+        metrics: Iterable[ComponentMetricId],
+    ) -> Self:  # pylint: disable=undefined-variable:
+        """Create instance of this class.
+
+        Subscribe for the requested component data and fetch only the latest component
+        metrics.
+
+        Args:
+            component_id: component id
+            metrics: metrics
+
+        Raises:
+            ValueError: If any requested metric id is not supported.
+
+        Returns:
+            This class instance
+        """
+        self: LatestMetricsFetcher[T] = await super().async_new(component_id, metrics)
+
+        for metric in metrics:
+            # pylint: disable=protected-access
+            if metric not in self._supported_metrics():
+                category = self._component_category()
+                raise ValueError(f"Metric {metric} not supported fro the {category}")
+
+        # pylint: disable=protected-access
+        self._receiver = await self._subscribe()
+        return self
+
+    async def fetch_next(self) -> Optional[ComponentMetricsData]:
+        """Fetch the latest component metrics.
+
+        Returns:
+            Component metrics data.
+            None if the channel was closed and fetching next element is impossible.
+        """
+        try:
+            data = await self._receiver.receive()
+        except ChannelClosedError as err:
+            _logger.error(
+                "Channel for component %d was closed. Err: %s",
+                self._component_id,
+                str(err),
+            )
+            return None
+
+        metrics = {}
+        for mid in self._metrics:
+            value = self._extract_metric(data, mid)
+            # There is no guarantee that all fields in component message are populated
+            if not math.isnan(value):
+                metrics[mid] = value
+
+        return ComponentMetricsData(self._component_id, data.timestamp, metrics)
+
+    @abstractmethod
+    def _extract_metric(self, data: T, mid: ComponentMetricId) -> float:
+        pass
+
+    @abstractmethod
+    def _supported_metrics(self) -> Set[ComponentMetricId]:
+        pass
+
+    @abstractmethod
+    def _component_category(self) -> ComponentCategory:
+        pass
```

Comment:
Funny that you don't get a complain from `mypy` here. Usually `pass` means that you have an implementation that does nothing, so it should complain about not returning a `float`. For abstract methods you should use `...` instead.
```suggestion
    @abstractmethod
    def _extract_metric(self, data: T, mid: ComponentMetricId) -> float:
        ...

    @abstractmethod
    def _supported_metrics(self) -> Set[ComponentMetricId]:
        ...

    @abstractmethod
    def _component_category(self) -> ComponentCategory:
        ...
```

Reply:
Thanks, fixed.